lib/BuildSystem/LaneBasedExecutionQueue.cpp

```diff
@@ -134,6 +159,22 @@ class LaneBasedExecutionQueue : public BuildExecutionQueue {
     readyJobsCondition.notify_one();
   }
 
+  virtual void cancelAllJobs() override {
+    // We have already been cancelled
+    if (killAfterTimeoutThread != nullptr) {
+      return;
+    }
+
+    {
+      std::unique_lock<std::mutex> lock(readyJobsMutex);
+
+      readyJobs.clear();
+    }
+
+    sendSignalToProcesses(SIGINT);
+    killAfterTimeoutThread = new std::thread(&LaneBasedExecutionQueue::killAfterTimeout, this);
```

Comment:
I think what should probably happen is that there should be a (`std::atomic`) flag which tracks the state of the queue, whether or not it is cancelling. If it is cancelling, then each lane should discard all subsequent work, and attempts to add new work should return an error (eventually, maybe no need to add the error handling part now).